README.md
# fhreads

## A PHP extension to get real system threading functionality in php userland.

[![Build Status](https://travis-ci.org/appserver-io-php/fhreads.svg?branch=master)](https://travis-ci.org/appserver-io-php/fhreads)

## What is it?

The extension gives you access to plain system threading functions for being able to implement threading utilites as part of the language directly in the PHP userland e.g. an abstract Thread class.

## How to use?

> This extension is in early-stage development. Please use with caution.

This is a simple example how to use a shared data object as normal php object with a thread safe counter property.
```php
<?php

require_once __DIR__ . DIRECTORY_SEPARATOR . "Thread.php";

// define counter thread class which counts the shared data objects counter property
class CounterThread extends Thread
{
    /**
     * Constructor which refs the shared data object to this
     * 
     * @param object $data
     */
    public function __construct($data)
    {
        $this->data = $data;
    }
    
    public function run()
    {
        fhread_mutex_lock($this->data->mutex);
        ++$this->data->counter;
        $this->data->{$this->getThreadId()} = $this->getThreadId();
        fhread_mutex_unlock($this->data->mutex);
    }
}

$data = new \stdClass();
$data->mutex = fhread_mutex_init();
$data->counter = 0;

$ths = array();
$tMax = 10;

// initiate threads
for ($i = 1; $i <= $tMax; $i++) {
    $ths[$i] = new CounterThread($data);
}
// start threads
for ($i = 1; $i <= $tMax; $i++) {
    $ths[$i]->start();
}
// wait for all thread to be finished by joining them
for ($i = 1; $i <= $tMax; $i++) {
    $ths[$i]->join();
}

// echo status
echo "$tMax threads finished..." . PHP_EOL;
// dump shared data object
var_dump($data);
// echo status
echo PHP_EOL . "finished script!" . PHP_EOL;
```

This should give us an amazing output! :)
```bash
10 threads finished...
object(stdClass)#1 (12) {
  ["mutex"]=>
  int(19463520)
  ["counter"]=>
  int(10)
  ["139792551667456"]=>
  int(139792551667456)
  ["139792533702400"]=>
  int(139792533702400)
  ["139792506754816"]=>
  int(139792506754816)
  ["139792488789760"]=>
  int(139792488789760)
  ["139792515737344"]=>
  int(139792515737344)
  ["139792524719872"]=>
  int(139792524719872)
  ["139792479807232"]=>
  int(139792479807232)
  ["139792560649984"]=>
  int(139792560649984)
  ["139792542684928"]=>
  int(139792542684928)
  ["139792497772288"]=>
  int(139792497772288)
}

finished script!
```

See `examples` folder for more usage demonstration.

More documentation and examples comming...



CHANGELOG.md
# Version 0.1.0

## Bugfixes

* None

## Features

* Initial release


fhreads.c
/*
  +----------------------------------------------------------------------+
  | PHP Version 5                                                        |
  +----------------------------------------------------------------------+
  | Copyright (c) 1997-2015 The PHP Group                                |
  +----------------------------------------------------------------------+
  | This source file is subject to version 3.01 of the PHP license,      |
  | that is bundled with this package in the file LICENSE, and is        |
  | available through the world-wide-web at the following url:           |
  | http://www.php.net/license/3_01.txt                                  |
  | If you did not receive a copy of the PHP license and are unable to   |
  | obtain it through the world-wide-web, please send a note to          |
  | license@php.net so we can mail you a copy immediately.               |
  +----------------------------------------------------------------------+
  | Author: Johann Zelger <jz@appserver.io>                              |
  +----------------------------------------------------------------------+
*/

/* $Id$ */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include "php.h"
#include "php_ini.h"
#include "ext/standard/info.h"
#include "php_fhreads.h"

ZEND_DECLARE_MODULE_GLOBALS(fhreads)

/* True global resources - no need for thread safety here */
void ***fhread_global_tsrm_ls;

static int le_fhreads;
static zend_object_handlers fhreads_handlers;
static zend_object_handlers *fhreads_zend_handlers;
static HashTable fhreads_objects;

pthread_mutex_t fhread_global_mutex = PTHREAD_MUTEX_INITIALIZER;

zend_uint fhread_global_objects_store_top = 1;

/* {{{ PHP_INI
 */
/* Remove comments and fill if you need to have entries in php.ini
PHP_INI_BEGIN()
    STD_PHP_INI_ENTRY("fhreads.global_value",      "42", PHP_INI_ALL, OnUpdateLong, global_value, zend_fhreads_globals, fhreads_globals)
    STD_PHP_INI_ENTRY("fhreads.global_string", "foobar", PHP_INI_ALL, OnUpdateString, global_string, zend_fhreads_globals, fhreads_globals)
PHP_INI_END()
*/
/* }}} */

void fhread_write_property(zval *object, zval *member, zval *value, const zend_literal *key TSRMLS_DC)
{
	zend_std_write_property(object, member, value, key TSRMLS_CC);
}

/* {{{ */
void fhreads_global_free(FHREAD *fhread) {
	// efree(&fhread);
} /* }}} */

/* {{{ */
static int fhread_opz_new_handler(ZEND_OPCODE_HANDLER_ARGS)
{
	// fhreaded objects to sync
	FHREAD *fhread;
	zend_mm_heap *old_mm_heap, *mm_heap;

	// lock global mutex
	pthread_mutex_lock(&fhread_global_mutex);

	// increment global object top counter
	FHREADS_EG(fhread_global_tsrm_ls, objects_store).top = fhread_global_objects_store_top++;

	if (fhread_global_objects_store_top == FHREADS_EG(fhread_global_tsrm_ls, objects_store).size) {

		FHREADS_EG(fhread_global_tsrm_ls, objects_store).size <<= 1;

		// set heap to be in global context scope
		mm_heap = zend_mm_set_heap(NULL, fhread_global_tsrm_ls);
		zend_mm_set_heap(mm_heap, fhread_global_tsrm_ls);
		old_mm_heap = zend_mm_set_heap(mm_heap TSRMLS_CC);

		// realloc new objects_bucket
		FHREADS_EG(fhread_global_tsrm_ls, objects_store).object_buckets = (zend_object_store_bucket *) erealloc(FHREADS_EG(fhread_global_tsrm_ls, objects_store).object_buckets, FHREADS_EG(fhread_global_tsrm_ls, objects_store).size * sizeof(zend_object_store_bucket));

		// reset heap to be back in current scope
		zend_mm_set_heap(old_mm_heap TSRMLS_CC);
	}

	// synchronize context fhread objects_store
	for (zend_hash_internal_pointer_reset(&fhreads_objects);
	     zend_hash_get_current_data(&fhreads_objects, (void **) &fhread) == SUCCESS;
	     zend_hash_move_forward(&fhreads_objects)
	) {
		FHREADS_EG(fhread->tsrm_ls, objects_store).free_list_head = FHREADS_EG(fhread_global_tsrm_ls, objects_store).free_list_head;
		FHREADS_EG(fhread->tsrm_ls, objects_store).top = fhread_global_objects_store_top;
		FHREADS_EG(fhread->tsrm_ls, objects_store).size = FHREADS_EG(fhread_global_tsrm_ls, objects_store).size;
		FHREADS_EG(fhread->tsrm_ls, objects_store).object_buckets = FHREADS_EG(fhread_global_tsrm_ls, objects_store).object_buckets;
	}

	// lock global mutex
	pthread_mutex_unlock(&fhread_global_mutex);

	// go on
	return ZEND_USER_OPCODE_DISPATCH;
}

/* {{{ Initialises the executor in fhreads context */
void fhread_init_executor(FHREAD *fhread TSRMLS_DC) /* {{{ */
{
	// check if executor was not inited before
	if (fhread->executor_inited == 0) {
		zend_init_fpu(TSRMLS_C);

		INIT_ZVAL(EG(uninitialized_zval));
		/* trick to make uninitialized_zval never be modified, passed by ref, etc. */
		Z_ADDREF(EG(uninitialized_zval));
		INIT_ZVAL(EG(error_zval));
		EG(uninitialized_zval_ptr)=&EG(uninitialized_zval);
		EG(error_zval_ptr)=&EG(error_zval);
/* destroys stack frame, therefore makes core dumps worthless */
#if 0&&ZEND_DEBUG
	original_sigsegv_handler = signal(SIGSEGV, zend_handle_sigsegv);
#endif

		EG(return_value_ptr_ptr) = NULL;
		EG(symtable_cache_ptr) = EG(symtable_cache) - 1;
		EG(symtable_cache_limit) = EG(symtable_cache) + SYMTABLE_CACHE_SIZE - 1;
		EG(no_extensions) = 0;
		EG(in_execution) = 0;
		EG(in_autoload) = NULL;
		EG(autoload_func) = NULL;
		EG(error_handling) = EH_NORMAL;

		// link functions
		EG(function_table) = FHREADS_CG(fhread->c_tsrm_ls, function_table);
		// link classes
		EG(class_table) = FHREADS_CG(fhread->c_tsrm_ls, class_table);
		// link constants
		EG(zend_constants) = FHREADS_EG(fhread->c_tsrm_ls, zend_constants);
		// link regular list
		EG(regular_list) = FHREADS_EG(fhread->c_tsrm_ls, regular_list);
		// link objects_store
		EG(objects_store) = FHREADS_EG(fhread->c_tsrm_ls, objects_store);

		zend_vm_stack_init(TSRMLS_C);
		zend_vm_stack_push((void *) NULL TSRMLS_CC);

		EG(ticks_count) = 0;
		EG(user_error_handler) = NULL;
		EG(current_execute_data) = NULL;

		EG(full_tables_cleanup) = 0;
	#ifdef ZEND_WIN32
		EG(timed_out) = 0;
	#endif

		EG(exception) = NULL;
		EG(prev_exception) = NULL;
		EG(scope) = NULL;
		EG(called_scope) = NULL;
		EG(This) = NULL;
		EG(active_op_array) = NULL;
		EG(active) = 1;
		EG(start_op) = NULL;

		// set executor inited flag
		fhread->executor_inited = 1;
	}

	// init executor globals for function run to execute
	EG(This) = (*fhread->runnable);
	EG(active_op_array) = (zend_op_array*) fhread->run;
	EG(scope) = Z_OBJCE_PP(fhread->runnable);
	EG(called_scope) = EG(scope);
	EG(in_execution) = 1;
}
/* }}} */

/* {{{ The routine to call for pthread_create */
void *fhread_routine (void *arg)
{
	// passed the object as argument
	FHREAD* fhread = (FHREAD *) arg;
	// init threadsafe manager local storage and create new context
	TSRMLS_D = fhread->tsrm_ls;
	// set interpreter context
	tsrm_set_interpreter_context(TSRMLS_C);
	// link server context
	SG(server_context) = FHREADS_SG(fhread->c_tsrm_ls, server_context);
	// init executor
	fhread_init_executor(fhread TSRMLS_CC);
	// exec run method
	zend_execute((zend_op_array*) fhread->run TSRMLS_CC);
	// exit thread
	pthread_exit(NULL);
#ifdef _WIN32
	return NULL; /* silence MSVC compiler */
#endif
} /* }}} */

/* {{{ proto fhread_tls_get_id()
	Obtain the identifier of currents ls context */
PHP_FUNCTION(fhread_tls_get_id)
{
	ZVAL_LONG(return_value, (long)TSRMLS_C);
} /* }}} */

/* {{{ proto fhread_mutex_init() */
PHP_FUNCTION(fhread_mutex_init)
{
	pthread_mutex_t *mutex;
	if ((mutex=(pthread_mutex_t*) calloc(1, sizeof(pthread_mutex_t)))!=NULL) {
		RETURN_LONG((ulong)mutex);
	}
} /* }}} */


/* {{{ proto fhread_mutex_lock() */
PHP_FUNCTION(fhread_mutex_lock)
{
	pthread_mutex_t *mutex;
	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "l", &mutex)==SUCCESS && mutex) {
		pthread_mutex_lock(mutex);
	}
} /* }}} */


/* {{{ proto fhread_mutex_unlock() */
PHP_FUNCTION(fhread_mutex_unlock)
{
	pthread_mutex_t *mutex;
	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "l", &mutex)==SUCCESS && mutex) {
		pthread_mutex_unlock(mutex);
	}
} /* }}} */

/* {{{ proto fhread_object_get_handle(object obj)
	Obtain the identifier of the given objects handle */
PHP_FUNCTION(fhread_object_get_handle)
{
	zval *obj;
	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "z", &obj) == FAILURE) {
		RETURN_NULL();
	}
	ZVAL_LONG(return_value, Z_OBJ_HANDLE_P(obj));
} /* }}} */

/* {{{ proto fhread_self()
	Obtain the identifier of the current thread. */
PHP_FUNCTION(fhread_self)
{
	ZVAL_LONG(return_value, tsrm_thread_id());
} /* }}} */

/* {{{ proto fhread_free_interpreter_context(long thread_id)
	frees an interpreter context created by a thread */
PHP_FUNCTION(fhread_free_interpreter_context)
{
	THREAD_T thread_id;
	int status;

	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "l", &thread_id) == FAILURE) {
		RETURN_NULL();
	}

	//void ***f_tsrm_ls = (void ***) ts_resource_ex(0, &thread_id);
	//printf("fhread_free_interpreter_context: %d\n", f_tsrm_ls);
	// tsrm_free_interpreter_context(f_tsrm_ls);
} /* }}} */

/* {{{ proto fhread_create()
   Create a new thread, starting with execution of START-ROUTINE
   getting passed ARG.  Creation attributed come from ATTR.  The new
   handle is stored in thread_id which will be returned to php userland. */
PHP_FUNCTION(fhread_create)
{
	char *gid;
	int gid_len;
	int status;
	FHREAD* fhread;

	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &gid, &gid_len) == FAILURE) {
		RETURN_NULL();
	}

	// check if fhread was alreade prepared for gid
	if (zend_hash_find(&fhreads_objects, gid, gid_len + 1, (void **)&fhread)  == FAILURE) {
		// prepare fhread
		fhread = (FHREAD *)malloc(sizeof(FHREAD));
		// init executor flag
		fhread->executor_inited = 0;
		// set creator tsrm ls
		fhread->c_tsrm_ls = fhread_global_tsrm_ls;
		// create new tsrm ls for thread routine
		fhread->tsrm_ls = tsrm_new_interpreter_context();
		// add to globals hash
		zend_hash_add(&fhreads_objects, gid, gid_len + 1, (void*)fhread, sizeof(FHREAD), NULL);
	}

	// get runnable zval from creator symbol table
	zend_hash_find(&EG(symbol_table), gid, gid_len + 1, (void**)&fhread->runnable);
	// inject fhread handlers for runnable zval
	Z_OBJ_HT_PP(fhread->runnable) = &fhreads_handlers;
	// get run function from function table
	zend_hash_find(&Z_OBJCE_PP(fhread->runnable)->function_table, "run", sizeof("run"), (void**)&fhread->run);

	// create thread and start fhread__routine
	status = pthread_create(&fhread->thread_id, NULL, fhread_routine, fhread);

	// return thread id
	RETURN_LONG((long)fhread->thread_id);
} /* }}} */

/* {{{ proto fhread_join()
   Make calling thread wait for termination of the given thread id.  The
   exit status of the thread is stored in *THREAD_RETURN, if THREAD_RETURN
   is not NULL.*/
PHP_FUNCTION(fhread_join)
{
	long thread_id;
	int status;

	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "l", &thread_id) == FAILURE) {
		RETURN_NULL();
	}

	// join thread with given id
	status = pthread_join((pthread_t)thread_id, NULL);

	RETURN_LONG((long)status);
} /* }}} */

/* {{{ PHP_MINIT_FUNCTION
 */
PHP_MINIT_FUNCTION(fhreads)
{
	fhread_global_tsrm_ls = TSRMLS_C;

	zend_set_user_opcode_handler(ZEND_NEW, fhread_opz_new_handler);

	// Setup standard and fhreads object handlers
	fhreads_zend_handlers = zend_get_std_object_handlers();
	memcpy(&fhreads_handlers, fhreads_zend_handlers, sizeof(zend_object_handlers));

	// override object handlers
	fhreads_handlers.write_property = fhread_write_property;
	fhreads_handlers.get = NULL;
	fhreads_handlers.set = NULL;
	fhreads_handlers.get_property_ptr_ptr = NULL;

#if PHP_VERSION_ID > 50399
	/* when the gc runs, it will fetch properties, every time */
	/* so we pass in a dummy function to control memory usage */
	/* properties copied will be destroyed with the object */
	fhreads_handlers.get_gc = NULL;
#endif

	/* If you have INI entries, uncomment these lines 
	REGISTER_INI_ENTRIES();
	*/

	ZEND_INIT_MODULE_GLOBALS(fhreads, NULL, NULL);

	zend_hash_init(&FHREADS_G(fhreads), 64, NULL, (dtor_func_t) fhreads_global_free, 1);
	zend_hash_init(&fhreads_objects, 64, NULL, (dtor_func_t) fhreads_global_free, 1);

	return SUCCESS;
} /* }}} */

/* {{{ PHP_MSHUTDOWN_FUNCTION
 */
PHP_MSHUTDOWN_FUNCTION(fhreads)
{
	/* uncomment this line if you have INI entries
	UNREGISTER_INI_ENTRIES();
	*/

	zend_hash_destroy(&FHREADS_G(fhreads));

	return SUCCESS;
} /* }}} */

/* Remove if there's nothing to do at request start */
/* {{{ PHP_RINIT_FUNCTION
 */
PHP_RINIT_FUNCTION(fhreads)
{
	return SUCCESS;
} /* }}} */

/* Remove if there's nothing to do at request end */
/* {{{ PHP_RSHUTDOWN_FUNCTION
 */
PHP_RSHUTDOWN_FUNCTION(fhreads)
{
	return SUCCESS;
} /* }}} */

/* {{{ PHP_MINFO_FUNCTION
 */
PHP_MINFO_FUNCTION(fhreads)
{
	php_info_print_table_start();
	php_info_print_table_header(2, "fhreads support", "enabled");
	php_info_print_table_row(2, "fhreads version", PHP_FHREADS_VERSION);
	php_info_print_table_end();

	/* Remove comments if you have entries in php.ini
	DISPLAY_INI_ENTRIES();
	*/
} /* }}} */

/* {{{ fhreads_functions[]
 *
 * Every user visible function must have an entry in fhreads_functions[].
 */
const zend_function_entry fhreads_functions[] = {
	PHP_FE(fhread_tls_get_id, 				NULL)
	PHP_FE(fhread_object_get_handle, 		NULL)
	PHP_FE(fhread_free_interpreter_context, NULL)
	PHP_FE(fhread_self, 					NULL)
	PHP_FE(fhread_create, 					NULL)
	PHP_FE(fhread_join, 					NULL)
	PHP_FE(fhread_mutex_init, 				NULL)
	PHP_FE(fhread_mutex_lock, 				NULL)
	PHP_FE(fhread_mutex_unlock,				NULL)
	PHP_FE_END	/* Must be the last line in fhreads_functions[] */
}; /* }}} */

/* {{{ fhreads_module_entry
 */
zend_module_entry fhreads_module_entry = {
	STANDARD_MODULE_HEADER,
	"fhreads",
	fhreads_functions,
	PHP_MINIT(fhreads),
	PHP_MSHUTDOWN(fhreads),
	PHP_RINIT(fhreads),		/* Replace with NULL if there's nothing to do at request start */
	PHP_RSHUTDOWN(fhreads),	/* Replace with NULL if there's nothing to do at request end */
	PHP_MINFO(fhreads),
	PHP_FHREADS_VERSION,
	STANDARD_MODULE_PROPERTIES
}; /* }}} */

#ifdef COMPILE_DL_FHREADS
ZEND_GET_MODULE(fhreads)
#endif

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */


fhreads.php
<?php

if (!class_exists('\Thread')) {
    require_once "ext/fhreads/Thread.php";
}

class TestObject
{
    public function __construct() {
        echo __METHOD__ . PHP_EOL;
    }
    
    public function __destruct() {
        echo __METHOD__ . PHP_EOL;
    }
}

class TestThread extends Thread
{
    public function __construct($data)
    {
        $this->data = $data;
        $this->testObjArray = array();
    }
    
    public function run()
    {
        
        usleep(rand(10000,20000));
        
        fhread_mutex_lock($this->data->mutex);
        ++$this->data->counter;
        fhread_mutex_unlock($this->data->mutex);
        
        /*
        $this->data->test[] = $this->getThreadId();
        
        fhread_mutex_lock($GLOBALS["mutex"]);
        $this->data->testObjArray[] = new TestObject();
        fhread_mutex_unlock($GLOBALS["mutex"]);
        
        
        
        
        $this->testInt = 123;

        $this->testStr = 'test';

        $this->testFloat = 1.234;

        $this->testBool = true;

        $this->testArray = array('a' => 'b');

        
        /*
        $this->testObj = new stdClass();
        $this->testObjArray[] = new stdClass();
        $this->testObjArray[0]->test = 'test';
        */
/*
        $this->data->test = $this->getThreadId();
        
        usleep(rand(1000,20000));
        */
    }
}

/*
for ($i = 1; $i < 200; $i++) {
    var_dump(fhread_tsrm_new_interpreter_context());
}


sleep(3);

exit(0);

*/

/*
$t = new TestThread();
$t->outside = 'outside';
$t->start();
$t->join();
*/


$data = new \stdClass();
$data->test = array();
$data->counter = 0;
$data->mutex = fhread_mutex_init();



$ths = array();
$tMax = 10;

$ctxCount = 1;


for ($i = 1; $i <= $tMax; $i++) {
    $ths[$i] = new TestThread($data);
}
while(1) {
    

for ($i = 1; $i <= $tMax; $i++) {
    $ths[$i]->start();
}

for ($i = 1; $i <= $tMax; $i++) {
    $ths[$i]->join();
}

echo "$tMax threads finished, restarting..." . PHP_EOL;

var_dump($data);

sleep(1);

}

echo PHP_EOL . "finished script!" . PHP_EOL;



php_fhreads.h
/*
  +----------------------------------------------------------------------+
  | PHP Version 5                                                        |
  +----------------------------------------------------------------------+
  | Copyright (c) 1997-2015 The PHP Group                                |
  +----------------------------------------------------------------------+
  | This source file is subject to version 3.01 of the PHP license,      |
  | that is bundled with this package in the file LICENSE, and is        |
  | available through the world-wide-web at the following url:           |
  | http://www.php.net/license/3_01.txt                                  |
  | If you did not receive a copy of the PHP license and are unable to   |
  | obtain it through the world-wide-web, please send a note to          |
  | license@php.net so we can mail you a copy immediately.               |
  +----------------------------------------------------------------------+
  | Author: Johann Zelger <jz@appserver.io>                              |
  +----------------------------------------------------------------------+
*/

/* $Id$ */

#ifndef PHP_FHREADS_H
#define PHP_FHREADS_H

extern zend_module_entry fhreads_module_entry;
#define phpext_fhreads_ptr &fhreads_module_entry

#define PHP_FHREADS_VERSION "0.1.0"

#ifdef PHP_WIN32
#	define PHP_FHREADS_API __declspec(dllexport)
#elif defined(__GNUC__) && __GNUC__ >= 4
#	define PHP_FHREADS_API __attribute__ ((visibility("default")))
#else
#	define PHP_FHREADS_API
#endif

#include <stdio.h>
#ifndef _WIN32
#include <pthread.h>
#include <sys/time.h>
#include <signal.h>
#else
#include <pthread.h>
#include <signal.h>
#endif

#include <main/SAPI.h>
#include <Zend/zend.h>
#include <Zend/zend_API.h>
#include <Zend/zend_alloc.h>
#include <Zend/zend_operators.h>
#include <Zend/zend_closures.h>
#include <Zend/zend_compile.h>
#include <Zend/zend_execute.h>
#include <Zend/zend_exceptions.h>
#include <Zend/zend_extensions.h>
#include <Zend/zend_globals.h>
#include <Zend/zend_hash.h>
#include <Zend/zend_ts_hash.h>
#include <Zend/zend_interfaces.h>
#include <Zend/zend_list.h>
#include <Zend/zend_object_handlers.h>
#include <Zend/zend_variables.h>
#include <Zend/zend_ptr_stack.h>
#include <Zend/zend_constants.h>
#include <Zend/zend_closures.h>
#include <Zend/zend_generators.h>
#include <Zend/zend_vm.h>

#ifdef ZTS
#include "TSRM.h"
#endif

ZEND_BEGIN_MODULE_GLOBALS(fhreads)
	HashTable fhreads;
ZEND_END_MODULE_GLOBALS(fhreads)

typedef struct _fhread {
	pthread_t thread_id;
	void ***c_tsrm_ls;
	TSRMLS_D;
	zend_function *run;
	zval **runnable;
	zend_object_handle *handle;
	int executor_inited;
} FHREAD;

void fhread_init_executor(FHREAD *fhread TSRMLS_DC);

/* {{{ TSRM manipulation */
#define FHREADS_FETCH_ALL(ls, id, type) ((type) (*((void ***) ls))[TSRM_UNSHUFFLE_RSRC_ID(id)])
#define FHREADS_FETCH_CTX(ls, id, type, element) (((type) (*((void ***) ls))[TSRM_UNSHUFFLE_RSRC_ID(id)])->element)
#define FHREADS_CG(ls, v) FHREADS_FETCH_CTX(ls, compiler_globals_id, zend_compiler_globals*, v)
#define FHREADS_CG_ALL(ls) FHREADS_FETCH_ALL(ls, compiler_globals_id, zend_compiler_globals*)
#define FHREADS_EG(ls, v) FHREADS_FETCH_CTX(ls, executor_globals_id, zend_executor_globals*, v)
#define FHREADS_SG(ls, v) FHREADS_FETCH_CTX(ls, sapi_globals_id, sapi_globals_struct*, v)
#define FHREADS_EG_ALL(ls) FHREADS_FETCH_ALL(ls, executor_globals_id, zend_executor_globals*)
/* }}} */

#ifdef ZTS
#define FHREADS_G(v) TSRMG(fhreads_globals_id, zend_fhreads_globals *, v)
#else
#define FHREADS_G(v) (fhreads_globals.v)
#endif

#endif	/* PHP_FHREADS_H */


/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */


examples/Thread.php
<?php

/**
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 *
 * PHP version 5
 *
 * @author    Johann Zelger <jz@appserver.io>
 * @copyright 2015 TechDivision GmbH <info@appserver.io>
 * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 * @link      https://github.com/appserver-io-php/fhreads
 * @link      http://www.appserver.io
 */

/**
 * Interface Runnable
 * 
 * Simple interface for runnables
 *
 * @author    Johann Zelger <jz@appserver.io>
 * @copyright 2015 TechDivision GmbH <info@appserver.io>
 * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 * @link      https://github.com/appserver-io/fhreads
 * @link      http://www.appserver.io
 */
interface Runnable
{
    public function run();
}

/**
 * Abstract Class Thread
 *
 * Simple thread abstract class which implements runnable interface
 *
 * @author    Johann Zelger <jz@appserver.io>
 * @copyright 2015 TechDivision GmbH <info@appserver.io>
 * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 * @link      https://github.com/appserver-io/fhreads
 * @link      http://www.appserver.io
 */
abstract class Thread implements Runnable
{
    /**
     * Holds thread id if started
     * 
     * @var int
     */
    protected $threadId = null;
    
    /**
     * Abstract run function
     * 
     * @return void
     */
    abstract function run();
    
    /**
     * Internal function to get globals identifier string
     * 
     * @return string
     */
    public function getGlobalsIdentifier()
    {
        return get_class($this) . "#" . fhread_object_get_handle($this);
    }
    
    /**
     * Start method which will prepare, create and starts a thread
     * 
     * @return void
     */
    public function start()
    {
        // set ref to globals for thread to use it from
        $GLOBALS[$this->getGlobalsIdentifier()] = $this;

        // create, start thread and save thread id 
        $this->threadId = fhread_create($this->getGlobalsIdentifier());
    }
    
    /**
     * Destroyes the thread and remove from from globals
     * 
     * @return void
     */
    public function destroy()
    {
        unset($GLOBALS[$this->getGlobalsIdentifier()]);
    }
    
    /**
     * Returns the thread id
     * 
     * @return int
     */
    public function getThreadId()
    {
        return $this->threadId;
    }
    
    /**
     * Joins the current thread by its thread id.
     * 
     * @return void
     */
    public function join()
    {
        if ($this->getThreadId()) {
            fhread_join($this->getThreadId());
        }
    }
}


examples/sharing.php
<?php

// to be compatible with pthreads lib
if (!class_exists('\Thread')) {
    require_once __DIR__ . DIRECTORY_SEPARATOR . "Thread.php";
}

// define counter thread class which counts the shared data objects counter property
class CounterThread extends Thread
{
    /**
     * Constructor which refs the shared data object to this
     * 
     * @param object $data
     */
    public function __construct($data)
    {
        $this->data = $data;
    }
    
    public function run()
    {
        // wait randomized time
        usleep(rand(10000,20000));
        // lock data object
        fhread_mutex_lock($this->data->mutex);
        // inc counter
        ++$this->data->counter;
        // add message that this thread was here
        $this->data->{$this->getThreadId()} = $this->getThreadId();
        // unlock data object
        fhread_mutex_unlock($this->data->mutex);
    }
}

// create data storage object
$data = new \stdClass();
// init mutex
$data->mutex = fhread_mutex_init();
// init counter property to be zero
$data->counter = 0;


// init threads array
$ths = array();
// define max threads
$tMax = 10;

// initiate threads
for ($i = 1; $i <= $tMax; $i++) {
    $ths[$i] = new CounterThread($data);
}

// start threads
for ($i = 1; $i <= $tMax; $i++) {
    $ths[$i]->start();
}

// wait for all thread to be finished by joining them
for ($i = 1; $i <= $tMax; $i++) {
    $ths[$i]->join();
}

// echo status
echo "$tMax threads finished..." . PHP_EOL;

// dump shared data object
var_dump($data);

// echo status
echo PHP_EOL . "finished script!" . PHP_EOL;

